File: behavioral/visitor.php
<?php
interface Visitor {
    public function visitCircle(Circle $c);
    public function visitSquare(Square $s);
}

interface Shape {
    public function accept(Visitor $v);
}

class Circle implements Shape {
    public function accept(Visitor $v){ $v->visitCircle($this); }
}
class Square implements Shape {
    public function accept(Visitor $v){ $v->visitSquare($this); }
}

class AreaVisitor implements Visitor {
    public function visitCircle(Circle $c){ echo "Circle area calculated\n"; }
    public function visitSquare(Square $s){ echo "Square area calculated\n"; }
}

// استفاده
$shapes = [new Circle(), new Square()];
$v = new AreaVisitor();
foreach($shapes as $s){ $s->accept($v); }

File: behavioral/visitor.go
package main

import "fmt"

type Visitor interface {
	VisitCircle(Circle)
	VisitSquare(Square)
}

type Shape interface{ Accept(Visitor) }

type Circle struct{}

func (c Circle) Accept(v Visitor) { v.VisitCircle(c) }

type Square struct{}

func (s Square) Accept(v Visitor) { v.VisitSquare(s) }

type AreaVisitor struct{}

func (AreaVisitor) VisitCircle(c Circle) { fmt.Println("Circle area calculated") }
func (AreaVisitor) VisitSquare(s Square) { fmt.Println("Square area calculated") }

func main() {
	shapes := []Shape{Circle{}, Square{}}
	v := AreaVisitor{}
	for _, s := range shapes {
		s.Accept(v)
	}
}
